quinn-proto/src/address_discovery.rs

```diff
@@ -0,0 +1,126 @@
+//! Address discovery types from
+//! <https://datatracker.ietf.org/doc/draft-seemann-quic-address-discovery/>
+
+use crate::VarInt;
+
+/// The role of each participant.
+///
+/// When enabled, this is reported as a transport parameter.
+#[derive(PartialEq, Eq, Clone, Copy, Debug, Default)]
+pub(crate) enum Role {
+    /// Is able to report observer addresses to other peers, but it's not interested in receiving
+    /// reports about its own address.
+    SendOnly,
+    /// Is interested on reports about its own observed address, but will not report back to other
+    /// peers.
+    ReceiveOnly,
+    /// Will both report and receive reports of observed addresses.
+    Both,
+    /// Address discovery is disabled.
+    #[default]
+    Disabled,
+}
+
+impl TryFrom<VarInt> for Role {
+    type Error = crate::transport_parameters::Error;
+
+    fn try_from(value: VarInt) -> Result<Self, Self::Error> {
+        match value.0 {
+            0 => Ok(Self::SendOnly),
+            1 => Ok(Self::ReceiveOnly),
+            2 => Ok(Self::Both),
+            _ => Err(crate::transport_parameters::Error::IllegalValue),
+        }
+    }
+}
+
+impl Role {
+    /// Whether address discovery is disabled.
+    pub(crate) fn is_disabled(&self) -> bool {
+        matches!(self, Self::Disabled)
+    }
+
+    /// Whether this peer's role allows for address reporting to other peers.
+    fn is_reporter(&self) -> bool {
+        matches!(self, Self::SendOnly | Self::Both)
+    }
+
+    /// Whether this peer's role accepts observed address reports.
+    fn receives_reports(&self) -> bool {
+        matches!(self, Self::ReceiveOnly | Self::Both)
+    }
+
+    /// Whether this peer should report observed addresses to the other peer.
+    pub(crate) fn should_report(&self, other: &Self) -> bool {
+        self.is_reporter() && other.receives_reports()
+    }
```

Comment:
I think maybe this should just be inlined into its call sites, I don't think the abstraction is helping